source/simulator/helper_functions.cc

```diff
@@ -289,19 +289,86 @@ namespace aspect
     auto write_statistics
       = [statistics_copy_ptr,this]()
     {
-      // Write into a temporary file for now so that we don't
-      // interrupt anyone who might want to look at the real
-      // statistics file while the program is still running
-      const std::string stat_file_name = parameters.output_directory + "statistics";
-      const std::string tmp_file_name = stat_file_name + ".tmp";
-
-      std::ofstream stat_file (tmp_file_name.c_str());
-      statistics_copy_ptr->write_text (stat_file,
+      // First write everything into a string in memory
+      std::ostringstream stream;
+      statistics_copy_ptr->write_text (stream,
                                        TableHandler::table_with_separate_column_description);
-      stat_file.close();
+      stream.flush();
+
+      const std::string statistics_contents = stream.str();
+
+      // Next find out whether we need to write everything into
+      // the statistics file, or whether it is enough to just write
+      // the last few bytes that were added since we wrote to that
+      // file again. The way we do that is by checking whether the
+      // first few bytes of the string we just created match what we
+      // had previously written. One might think that they never should,
+      // but the statistics object automatically sizes the column widths
+      // of its output to match what is being written, and so if a later
+      // entry requires more width, then even the first columns are
+      // changed -- in that case, we will have to write everything,
+      // not just append one line.
+      const bool write_everything
+        = ( // We may have never written anything. More precisely, this
+            // case happens if the statistics_last_write_size is at the
+            // value initialized by the Simulator::Simulator()
+            // constructor, and this can happen in two situations:
+            // (i) At the end of the first time step; and (ii) at the
+            // end of the first time step after restart since the
+            // variable we query here is not serialized. It is clear that
+            // at the end of the first time step, we want to write the
+            // entire contents of the statistics object. For the second
+            // case, this is also appropriate since someone may have
+            // previously restarted from a checkpoint, run a couple of
```

Comment:
This also happens without a "previous restart" if the simulation continued after writing a checkpoint (very likely).